### ISSUE TITLE
Add notes to each split category & docked edit transaction panel

### ISSUE THREAD

**Issue by jrandhawa09** [moneymanagerex/moneymanagerex#4106]
As someone trying to transition from MS Money (yes, I know really old software but does the job for me!), it would be nice if the software allowed us to:

1. add separate Note text to each split category and include these in reports/exports
2. have the option for "edit transaction" panel to be docked in bottom (or top) of the main window transactions rather than as a popup (maybe already exists but couldn't see the options to enable it)

Thanks

**Comment by vomikan** [comment]
#910

**Comment by whalley** [comment]
> "add separate Note text to each split category and include these in reports/exports"

Implemented (see https://github.com/moneymanagerex/moneymanagerex/issues/910)

Changing name of this to reflect docked edit.

**Comment by tactilis** [comment]
@jrandhawa09's request for a docked Edit Transaction panel probably seemed a relatively simple change - but, having just transitioned from MS Money, it is possible that they were unaware that there are _multiple_ Transaction dialogs in MMEX, each having different layouts and sizes. This would make 'docking' much more of a redesign task than perhaps they expected.

For example, all of these variations would need to be accommodated in a docked panel:
 
![image](https://user-images.githubusercontent.com/4060092/197739804-cb24e671-1945-488d-9f34-e7f2709ada5b.png)
 
![image](https://user-images.githubusercontent.com/4060092/197739921-895a32dd-dca9-4b74-be11-be1ab308bc9f.png)
 
![image](https://user-images.githubusercontent.com/4060092/197740031-7fb4ad74-34a9-4242-810f-d24d6cdc5e40.png)
 
![image](https://user-images.githubusercontent.com/4060092/197740094-556fa545-a31d-49ab-8569-79956ca27125.png)
 
![image](https://user-images.githubusercontent.com/4060092/197739980-76f50ce7-a972-4d2e-9f18-0d2576580ab0.png)
  
Instead of a docking panel, I would much prefer that the Transaction dialogs simply remembered the position that the user places them in and have them re-open there instead of always opening in the centre of the transaction list as happens currently:
 
https://user-images.githubusercontent.com/4060092/197752099-4086f274-9110-4a92-90b8-7eb5a939fae5.mp4
 
Some dialogs, such as Transaction Filter, do this already (although there is currently a bug #5215 that sometimes makes them 'forget' their saved position).

Once that bug is resolved, could the remembering of the user's chosen positions be applied to the Transaction dialogs please?

**Comment by whalley** [comment]
> Instead of a docking panel, I would much prefer that the Transaction dialogs simply remembered the position that the user places them in and have them re-open there instead of always opening in the centre of the transaction list as happens currently:

That too is not as simple as it seems. An attempt to save dialog position was attempted here: https://github.com/moneymanagerex/moneymanagerex/issues/4237 it gets messy if monitor configuration is changed or application is re-opened/moved on screens/monitors and closed without opening transaction dialog.  You the starting thinking about saving relative positions of main app and dialog, but that doesn't work either because of what has been discussed.

**Comment by tactilis** [comment]
@whalley 
> An attempt to save dialog position was attempted here: https://github.com/moneymanagerex/moneymanagerex/issues/4237

Yes, I did remember that problem but it feels as though we are sacrificing the benefit for the many of being able to place dialogs where they want them and have them stay there _within a session_ (at least) against trying to also satisfy the fewer number of people* who change monitor configuration or re-open/move the application to different screens/monitors.

* Yes, I know that includes you Mark😄 

As a partial workaround, I wondered whether we could combine the resetting of dialog positions if they become 'lost' with the proposed Reset View enhancement in #3681?

Or am I not understanding the problem sufficiently?

**Comment by tactilis** [comment]
As a partial fix I'd be happy for the Edit Transaction dialog (and other dialogs) to always appear in their default location, i.e the centre of the transaction list, the first time the dialog is displayed within a session but then to remember the position the user moves it to and be displayed there when the dialog is re-opened _within the same session_.

**Comment by whalley** [comment]
> As a partial fix I'd be happy for the Edit Transaction dialog (and other dialogs) to always appear in their default location, i.e the centre of the transaction list, the first time the dialog is displayed within a session but then to remember the position the user moves it to and be displayed there when the dialog is re-opened _within the same session_.

If I implemented that people would then raise defect reports to say that the position is not remembered when the application is re-opened :grinning: It's a no-win! 😞

**Comment by tactilis** [comment]
> If I implemented that people would then raise defect reports to say that the position is not remembered when the application is re-opened 😀

Smiley noted.

But it would be better than the current situation where I frequently end up moving dialogs out of the way and cursing when next time I open them they are in the way again. And it would satisfy, to a degree, the motivation for the OP's request for dialogs to be docked so they don't obscure the transaction list.

**Comment by whalley** [comment]
What about adding a option...
Dialog opening location preference : Center / Left / Right / Bottom

E.g. if Right selected...
https://user-images.githubusercontent.com/487345/197846801-c6843a82-0259-43e6-9871-4e3d498a86ae.mp4

It should handle overflow correctly also by fitting the window if it would go off-screen.
https://user-images.githubusercontent.com/487345/197847273-e08e5aeb-4e13-443e-9ce8-e56a50526c5c.mp4

**Comment by tactilis** [comment]
> What about adding an option...
> Dialog opening location preference : Center / Left / Right / Bottom

That works well on your 25:10 ultrawide monitor.

This is what I have to do on my 1920 x 1080 laptop:
  
![image](https://user-images.githubusercontent.com/4060092/197864079-626b43a4-c2ce-4af9-9a92-43f0a5642155.png)
 
I park the dialog bottom-right so that I can see as much of the transaction list as possible.

In general, adding more and more options is a bad thing. If possible, it's better to try to make the application "_Just Do The Right Thing_".

Every choice added means a bit more thinking time when a user looks at the Options and more legacy cruft to support in the future. It's often hard to take options away.

**Comment by whalley** [comment]
> In general, adding more and more options is a bad thing. If possible, it's better to try to make the application "Just Do The Right Thing".

Of course that assumes "The Right Thing" is the same for everyone......

**Comment by tactilis** [comment]
> Of course that assumes "The Right Thing" is the same for everyone...

Agreed that it's not _always_ the exactly same for _all_ - but then we need to look at likely numbers who would benefit from a particular solution:

- I think that everyone would benefit from dialogs staying where the user chooses to put them, whatever the mechanism for placing them there.

- I think a lesser number would benefit from a more restrictive Center / Left / Right / Bottom option.

The current behaviour is (for me) the worst. I'm constantly having to reposition dialogs when I open them. It's rather frustrating.

**Comment by jrandhawa09** [comment]
Hi 

Just my 2 cents worth. When I requested this I had something similar to this mockup in mind rather than floating panels which could be docked (sorry about the dodgy picture but should give you an idea as to what I was looking for). I don't mind if the splits etc appear as a popup in the middle but the main entries should be at the bottom (or even top)

![Test_mockup](https://user-images.githubusercontent.com/20520383/197898697-19fc2b45-c8f9-42f8-8d65-8704991c2045.jpg)

**Comment by whalley** [comment]
@jrandhawa09 That is a great contribution to the discussion and can be achieved, I quite like it. I could actually complement the existing dialog as a 'quick entry' mechanism.

**Comment by tactilis** [comment]
Looks good @jrandhawa09 

@whalley
> ...and can be achieved

That's good to hear. I had been concerned about the amount of work that would be required to accommodate all the dialog variants shown: https://github.com/moneymanagerex/moneymanagerex/issues/4106#issuecomment-1290350994.

**Dialog Modality**
If this enhancement goes ahead, then we need to consider the fact that the current transaction dialogs are all modal and are only displayed for viewing/editing when the user double clicks on a transaction row or right-clicks and selects `Edit Transaction(s)...`. No other interaction within MMEX is possible while the dialog is displayed.

Should we preserve this behaviour, i.e displaying the transaction's details in the panel in a modal fashion, which means that the user will have to click OK or Cancel to 'close' the details in the panel before they are able to click on another row (or any other part of the MMEX UI).

Or, should the contents of the panel be non-modal and be updated whenever a transaction row is highlighted as a means of quickly viewing the full transaction details (eg the full Notes, all the Custom Fields). Clicking (i.e. highlighting) another transaction row would update the panel with that row's details.

If the user makes any edits on the panel then the OK button becomes active and can be clicked to apply the changes.

Highlighting multiple rows would operate in a similar way to the `Edit Transactions...` Multi Transactions Update functionality but with the details being displayed (and edited) in the panel.

Non-modal behaviour is more flexible for the user but would obviously be more complex to implement.

**Comment by whalley** [comment]
> I had been concerned about the amount of work that would be required to accommodate all the dialog variants shown

It would be a good opportunity to rationalise these, and introduce some commonality in the code.

> It would have to be non-modal.

Agreed.

**Comment by tactilis** [comment]
If you make the behaviour non-modal (which IMO is the best solution), then you'll have to protect against (or handle appropriately) the user making changes in other parts of MMEX which invalidates the information displayed in the transaction panel.

For example, the user could start editing the transaction in the panel and before OK'ing they could:
- `Edit Account...` and change the account name or the currency.
- Change the name or type of Custom Fields that are set in that transaction.
- `Relocate` the payee that is used in the transaction.
- etc

All solvable by preventing or handling but will need thinking about.

**Comment by dbolton** [comment]
My preference would be to allow editing directly in each row (similar to inline editing of a spreadsheet). The benefit is you see the text or number you want to edit, then you edit it in place. 

In the current pop-up window you have to find the text or number you were just looking at in a new part of the screen and a completely different layout. I found that disorienting when I first started using MMEX.

Regarding the non-modal question, we could save changes when the input loses focus. (Rather than forcing the user to click OK).
